File: 322-coin-change/322-coin-change.scala
object Solution {
    def coinChange(coins: Array[Int], amount: Int): Int = {
        val memo: Array[Int] = Array.fill(amount + 1)(-2)
        memo(0) = 0
        helper(amount, coins, memo)
    }
    
    def helper(amount: Int, coins: Array[Int], memo: Array[Int]): Int = {
        if (memo(amount) != -2) { memo(amount) } 
        else {
            var min = -1
            
            for(coin <- coins) {
                if (amount >= coin) {
                    val v = helper(amount - coin, coins, memo)
                    if (v != -1) {
                        if (min == -1) {
                            min = 1 + v
                        } else {
                            min = Math.min(1 + v, min)
                        }
                    }
                }
            }
            
            memo(amount) = min
            min
        }
    }
}

File: 322-coin-change/322-coin-change.java
class Solution {
    public int coinChange(int[] coins, int amount) {
        int[] dp = new int[amount + 1];
        
        Arrays.fill(dp, -1);
        dp[0] = 0;
        
        for(int i = 1; i <= amount; i++) {
            int min = Integer.MAX_VALUE;
            boolean found = false;
            
            for(int j = 0; j < coins.length; j++) {
                int coin = coins[j];
                if (i >= coin && dp[i - coin] != -1) {
                    min = Math.min(min, dp[i - coin] + 1);
                    found = true;
                }
            }
            
            if (found) dp[i] = min;
        }
        
        return dp[amount];
    }
}